File: engine/zeromiss/seeker.py
"""The seeker / sensor model.

A perfect seeker makes guidance look too easy and hides the real engineering. The
seeker turns *true* relative geometry into the *measured* LOS rate the guidance law
actually receives, degraded by:

* **Seeker lag** — a first-order lag (time constant ``tau``) on the LOS rate.
* **Noise** — range-independent angular jitter ``sigma_theta`` plus optional
  **glint** that worsens as ``R -> 0`` (the classic terminal-homing headache).
* **Field of view** — a ``+/- fov`` half-angle gimbal limit; exceed it and you get
  **loss-of-lock** (the reticle goes red and the missile coasts).
* **Finite update rate** — the seeker runs at e.g. 100 Hz under the 1 kHz physics with
  a zero-order hold, exposing sample-rate effects.

The contract::

    class Seeker:
        def measure(self, g: Geometry, dt: float, rng) -> SeekerOut: ...
"""

from __future__ import annotations

import math
from dataclasses import dataclass

from .frames import Geometry, deg2rad


@dataclass(frozen=True, slots=True)
class SeekerConfig:
    """Seeker tuning."""

    ideal: bool = True               # if True, perfect measurement (ignores all below)
    tau: float = 0.1                 # LOS-rate lag time constant [s]
    noise_mrad: float = 1.0          # range-independent angular noise, RMS [mrad]
    glint: bool = False              # enable range-dependent (1/R) glint noise
    glint_ref_m: float = 1000.0      # glint reference range: sigma_glint = noise at R=this
    glint_gain_mrad: float = 2.0     # glint magnitude [mrad] at glint_ref_m
    fov_deg: float = 30.0            # gimbal/FOV half-angle [deg]
    update_hz: float = 100.0         # seeker sample rate [Hz]
    loss_of_lock_timeout_s: float = 0.5  # coast this long out-of-FOV before giving up

    @property
    def fov_rad(self) -> float:
        return deg2rad(self.fov_deg)

    @property
    def update_dt(self) -> float:
        return 1.0 / self.update_hz if self.update_hz > 0 else 0.0


@dataclass(frozen=True, slots=True)
class SeekerOut:
    """One seeker measurement."""

    lambda_dot: float    # measured LOS rate [rad/s] (what guidance consumes)
    locked: bool         # currently tracking the target
    in_fov: bool         # target within the gimbal field of view this step
    sigma: float         # noise std actually applied this update [rad]
    lock_lost_for: float  # seconds spent out of lock (0 while locked)


class Seeker:
    """Stateful seeker filter. One instance per engagement.

    All randomness comes from the ``rng`` passed into :meth:`measure`,
    so a fixed seed reproduces the run. The ideal seeker uses no randomness, which is
    what the TypeScript cross-validation fixtures rely on.
    """

    def __init__(self, cfg: SeekerConfig | None = None) -> None:
        self.cfg = cfg or SeekerConfig()
        self._t_since_update = math.inf   # force an update on the first call
        self._held = SeekerOut(0.0, True, True, 0.0, 0.0)
        self._lock_lost_for = 0.0
        self._locked = True

    def reset(self) -> None:
        self._t_since_update = math.inf
        self._held = SeekerOut(0.0, True, True, 0.0, 0.0)
        self._lock_lost_for = 0.0
        self._locked = True

    def measure(self, g: Geometry, dt: float, rng) -> SeekerOut:
        """Return the measured LOS rate for this physics step.

        Parameters
        ----------
        g : Geometry
            True relative geometry this step.
        dt : float
            Physics step [s].
        rng : numpy.random.Generator
            The single seeded RNG threaded through the engagement.
        """
        cfg = self.cfg

        # --- Ideal path: perfect, instantaneous, unlimited FOV. ---
        if cfg.ideal:
            return SeekerOut(g.lambda_dot, True, True, 0.0, 0.0)

        # --- Field-of-view / loss-of-lock bookkeeping (runs every physics step). ---
        in_fov = g.los_off_boresight <= cfg.fov_rad
        if in_fov:
            self._lock_lost_for = 0.0
            self._locked = True
        else:
            self._lock_lost_for += dt
            if self._lock_lost_for > cfg.loss_of_lock_timeout_s:
                self._locked = False

        # --- Zero-order hold between seeker updates. ---
        self._t_since_update += dt
        if self._t_since_update < cfg.update_dt and self._held is not None:
            held = self._held
            return SeekerOut(
                held.lambda_dot if self._locked else 0.0,
                self._locked,
                in_fov,
                held.sigma,
                self._lock_lost_for,
            )

        Ts = max(self._t_since_update, cfg.update_dt)
        self._t_since_update = 0.0

        # --- Noise on the measured LOS rate. ---
        sigma_ang = cfg.noise_mrad * 1e-3  # mrad -> rad
        if cfg.glint and g.R > 1e-6:
            sigma_ang += (cfg.glint_gain_mrad * 1e-3) * (cfg.glint_ref_m / g.R)
        # Angular noise on the LOS angle turns into rate noise ~ sigma_ang / Ts.
        rate_noise = rng.normal(0.0, sigma_ang) / Ts if sigma_ang > 0 else 0.0
        measured = g.lambda_dot + rate_noise

        # --- First-order lag on the LOS rate. ---
        if cfg.tau > 0.0:
            alpha = 1.0 - math.exp(-Ts / cfg.tau)
        else:
            alpha = 1.0
        prev = self._held.lambda_dot
        filtered = prev + alpha * (measured - prev)

        out = SeekerOut(
            lambda_dot=filtered if self._locked else 0.0,
            locked=self._locked,
            in_fov=in_fov,
            sigma=sigma_ang,
            lock_lost_for=self._lock_lost_for,
        )
        self._held = out
        return out

File: web/lib/sim/seeker.ts
/**
 * Seeker / sensor model — port of `engine/zeromiss/seeker.py`.
 * Lag, noise, FOV, loss-of-lock, finite update rate (zero-order hold).
 */

import { deg2rad, type Geometry } from "./frames";
import type { RNG } from "./rng";

export interface SeekerConfig {
  ideal: boolean;
  tau: number;
  noiseMrad: number;
  glint: boolean;
  glintRefM: number;
  glintGainMrad: number;
  fovDeg: number;
  updateHz: number;
  lossOfLockTimeoutS: number;
}

export const defaultSeeker: SeekerConfig = {
  ideal: true,
  tau: 0.1,
  noiseMrad: 1.0,
  glint: false,
  glintRefM: 1000.0,
  glintGainMrad: 2.0,
  fovDeg: 30.0,
  updateHz: 100.0,
  lossOfLockTimeoutS: 0.5,
};

export interface SeekerOut {
  lambdaDot: number;
  locked: boolean;
  inFov: boolean;
  sigma: number;
  lockLostFor: number;
}

export class Seeker {
  cfg: SeekerConfig;
  private tSinceUpdate = Infinity;
  private heldLd = 0.0;
  private heldSigma = 0.0;
  private lockLostFor = 0.0;
  private locked = true;

  constructor(cfg: Partial<SeekerConfig> = {}) {
    this.cfg = { ...defaultSeeker, ...cfg };
  }

  reset(): void {
    this.tSinceUpdate = Infinity;
    this.heldLd = 0.0;
    this.heldSigma = 0.0;
    this.lockLostFor = 0.0;
    this.locked = true;
  }

  private get fovRad(): number {
    return deg2rad(this.cfg.fovDeg);
  }
  private get updateDt(): number {
    return this.cfg.updateHz > 0 ? 1.0 / this.cfg.updateHz : 0.0;
  }

  measure(g: Geometry, dt: number, rng: RNG | null): SeekerOut {
    const cfg = this.cfg;
    if (cfg.ideal) {
      return { lambdaDot: g.lambdaDot, locked: true, inFov: true, sigma: 0.0, lockLostFor: 0.0 };
    }

    const inFov = g.losOffBoresight <= this.fovRad;
    if (inFov) {
      this.lockLostFor = 0.0;
      this.locked = true;
    } else {
      this.lockLostFor += dt;
      if (this.lockLostFor > cfg.lossOfLockTimeoutS) this.locked = false;
    }

    this.tSinceUpdate += dt;
    if (this.tSinceUpdate < this.updateDt) {
      return {
        lambdaDot: this.locked ? this.heldLd : 0.0,
        locked: this.locked,
        inFov,
        sigma: this.heldSigma,
        lockLostFor: this.lockLostFor,
      };
    }

    const Ts = Math.max(this.tSinceUpdate, this.updateDt);
    this.tSinceUpdate = 0.0;

    let sigmaAng = cfg.noiseMrad * 1e-3;
    if (cfg.glint && g.R > 1e-6) {
      sigmaAng += cfg.glintGainMrad * 1e-3 * (cfg.glintRefM / g.R);
    }
    const rateNoise = sigmaAng > 0 && rng ? rng.normal(0.0, sigmaAng) / Ts : 0.0;
    const measured = g.lambdaDot + rateNoise;

    const alpha = cfg.tau > 0 ? 1.0 - Math.exp(-Ts / cfg.tau) : 1.0;
    this.heldLd = this.heldLd + alpha * (measured - this.heldLd);
    this.heldSigma = sigmaAng;

    return {
      lambdaDot: this.locked ? this.heldLd : 0.0,
      locked: this.locked,
      inFov,
      sigma: sigmaAng,
      lockLostFor: this.lockLostFor,
    };
  }
}
